sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala

```diff
@@ -130,460 +138,507 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
       val version = if (i == 0) "2.3.0" else "1.3.0"
       val funcCall = if (i == 0) "() => func" else "func"
       println(s"""
-        |/**
-        | * Register a user-defined function with ${i} arguments.
-        | * @since $version
-        | */
-        |def register(name: String, f: UDF$i[$extTypeArgs], returnType: DataType): Unit = {
-        |  val func = f$anyCast.call($anyParams)
-        |  def builder(e: Seq[Expression]) = if (e.length == $i) {
-        |    ScalaUDF($funcCall, returnType, e)
-        |  } else {
-        |    throw new AnalysisException("Invalid number of arguments for function " + name +
-        |      ". Expected: $i; Found: " + e.length)
-        |  }
-        |  functionRegistry.createOrReplaceTempFunction(name, builder)
-        |}""".stripMargin)
+         |/**
+         | * Registers a deterministic user-defined function with ${i} arguments.
+         | * @since $version
+         | */
+         |def register(name: String, f: UDF$i[$extTypeArgs], returnType: DataType): Unit = {
+         |  register(name, f, returnType, deterministic = true)
+         |}
+         |
+         |/**
+         | * Registers a user-defined function with ${i} arguments.
+         | * @since 2.3.0
+         | */
+         |def register(name: String, f: UDF$i[$extTypeArgs], returnType: DataType, deterministic: Boolean): Unit = {
```

Comment:
as we have to add new APIs, why not we add a bunch of `def udf(f: UDF1[_, _]): UserDefinedFunction` instead of a bunch of `def register(name: String, f: UDF$i[$extTypeArgs], returnType: DataType, deterministic: Boolean)`?